[UI-side compositing] Guard against UI process hangs caused by incorrect frame rate data

smfr opened the pull request:
#### 3f7af1859b92db6b242e50fa76450385e91628e0
<pre>
[UI-side compositing] Guard against UI process hangs caused by incorrect frame rate data
<a href="https://bugs.webkit.org/show_bug.cgi?id=256399">https://bugs.webkit.org/show_bug.cgi?id=256399</a>
rdar://108954321

Reviewed by Chris Dumez.

The spin report in the radar suggests that the Safari main thread is blocked on
RemoteLayerTreeEventDispatcher::mainThreadDisplayDidRefresh() because the scrolling thread
is stuck in RemoteLayerTreeEventDispatcher::waitForRenderingUpdateCompletionOrTimeout().
The only way this can happen is that we send a long timeout to m_stateCondition.waitUntil(),
but this timeout should be at most half of the frame duration.

So protect against bad m_scrollingTree-&gt;frameDuration() data by imposing a cap
at at most 32ms on this timeout, and release log if it seems wrong.

* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeEventDispatcher.cpp:
(WebKit::RemoteLayerTreeEventDispatcher::waitForRenderingUpdateCompletionOrTimeout):

Canonical link: <a href="https://commits.webkit.org/263806@main">https://commits.webkit.org/263806@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f0a555591c5ec443bce62fcbaccfa196d56660c1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/5635 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/5787 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/5977 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/7187 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/5624 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/6012 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/5760 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/7606 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/5738 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/5790 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/5049 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/7229 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/3267 "Passed tests") | | [💥 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/12179 "An unexpected error occured. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/5136 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/5149 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/6971 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/5585 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/4569 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/5035 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1372 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/9144 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/5396 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeEventDispatcher.cpp:
```diff
@@ -450,6 +450,14 @@ void RemoteLayerTreeEventDispatcher::waitForRenderingUpdateCompletionOrTimeout()
+    constexpr auto maximumTimeoutDelay = 32_ms;
+    auto maximumTimeoutTime = currentTime + maximumTimeoutDelay;
+    if (timeoutTime > maximumTimeoutTime) {
+        RELEASE_LOG(DisplayLink, "%p - [webPageID=%" PRIu64 "] RemoteLayerTreeEventDispatcher::waitForRenderingUpdateCompletionOrTimeout - bad timeout %.2fms into the future (frame duration %.2fms)", this, m_pageIdentifier.toUInt64(),
```
RELEASE_LOG_ERROR no?

cdumez reviewed (approved):
LGTM

smfr committed:
[UI-side compositing] Guard against UI process hangs caused by incorrect frame rate data
https://bugs.webkit.org/show_bug.cgi?id=256399
rdar://108954321

Reviewed by Chris Dumez.

The spin report in the radar suggests that the Safari main thread is blocked on
RemoteLayerTreeEventDispatcher::mainThreadDisplayDidRefresh() because the scrolling thread
is stuck in RemoteLayerTreeEventDispatcher::waitForRenderingUpdateCompletionOrTimeout().
The only way this can happen is that we send a long timeout to m_stateCondition.waitUntil(),
but this timeout should be at most half of the frame duration.

So protect against bad m_scrollingTree->frameDuration() data by imposing a cap
at at most 32ms on this timeout, and release log if it seems wrong.

* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeEventDispatcher.cpp:
(WebKit::RemoteLayerTreeEventDispatcher::waitForRenderingUpdateCompletionOrTimeout):

Canonical link: https://commits.webkit.org/263806@main
